File: lintcode/easy/insertion_sort_list/cpp/insertion_sort_list.cpp
/*
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/insertion-sort-list
@Language: C++
@Datetime: 17-02-15 23:13
*/

/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @return: The head of linked list.
     */
    ListNode *insertionSortList(ListNode *head) 
    {
        ListNode dummy{ 0 }, *src{ head };
            
        while (src != nullptr) {
            if (src->next != nullptr && src->val > src->next->val) {
                break;
            }
            src = src->next;
        }
        
        if (src == nullptr) {
            return head;
        }
            
        src = head;
            
        while (src != nullptr)
        {
            ListNode *next{ src->next };
            ListNode *dest{ &dummy };
            
            while (dest->next != nullptr && src->val > dest->next->val) {
                dest = dest->next;
            }
            
            src->next  = dest->next;
            dest->next = src;
            src        = next;
        }
        
        return dummy.next;
    }
};

File: lintcode/easy/insertion_sort_list/py/insertion_sort_list.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/insertion-sort-list
@Language: Python
@Datetime: 17-02-15 21:47
'''

"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of linked list.
    @return: The head of linked list.
    """ 
    def insertionSortList(self, head):
        dummy = ListNode(0)
        src   = head
        
        while src != None:
            if src.next != None and src.val > src.next.val:
                break
            
            src = src.next
        else:
            return head
        
        src = head
        
        while src != None:
            nxt = src.next
            dst = dummy
            
            while dst.next != None and dst.next.val < src.val:
                dst = dst.next
            
            src.next = dst.next
            dst.next = src
            src      = nxt
            
        return dummy.next